### ISSUE TITLE
syntax: support inline comments inside backquotes

### ISSUE THREAD

**Issue by improbable-nickkrempel** [mvdan/sh#636]
For long commands and pipelines split across multiple lines, a common workaround to add comments in the middle is to use the following syntax, slightly abusing subshells, but seems to be the cleanest way overall:

```
command -option value `# comment goes here` \
  -another-option another-value `# another comment goes here` \
  -xyz
```

This syntax seems to be explicitly supported by `shellcheck`, but `shfmt` either gets confused or replaces it with `$(...)` subshell syntax (which won't generally work properly in combination with `#`). This in turn triggers `shellcheck` warnings. I think you can use `$(: ...)` as a workaround, but then you also have to disable the `shellcheck` warning for an unquoted substitution (quoting it would result in extraneous empty parameters).

Would it be possible for `shfmt` to recognize this special comment syntax and leave it alone (possibly under the control of a command-line flag)?

**Comment by mvdan** [comment]
Have you seen https://github.com/mvdan/sh/issues/342?

**Comment by improbable-nickkrempel** [comment]
Thanks, I've read that and it looks applicable here. It doesn't resolve what to do with existing code that looks like this though:

```
  # Use aquery to get the list of output files of the `CreatePatch` action,
  # Then strip the patch path down to that of its source file, and apply
  # the patch file generated by Bazel to the original source file.
  # shellcheck disable=SC2016
  CLANG_FORMAT="$(clangFormatLocation)" runBazel aquery --config lint --include_aspects --output_groups clang_format_patches_only "mnemonic(\"CreatePatch\", $(generateAqueryTargetString))" --output textproto \
    `# Get relative paths to source files` \
    `# perl used instead of grep --perl-regexp since grep macOS doesnt support it` \
    | perl -ne "while(/(?<=exec_path: \"${PREFIX//\//\\/}\/).*\.patch_.+(?=\")/g){print \"\$&\n\";}" \
    `# Create the patch commands which, when executed patch the source files.` \
    `# --binary flag used to correctly handle line endings on Windows.` \
    | xargs -L1 -I 'PATH_TO_SOURCE_FILE' echo 'FILE_PATH="PATH_TO_SOURCE_FILE"; echo patch --binary "${FILE_PATH%.patch_*}" "'"${BAZEL_BIN}"'/${FILE_PATH}"' \
    `# De-duplicate the patch commands, such that there is at most one patch command for each source file.` \
    `# There are N patch files per source files, where N is the number of bazel targets directly including the target.` \
    `# The format of the commands being de-duplicated is: patch --binary -source file- -patch file-` \
     | sh `# Calls the echo command generated by xargs, resulting in the patch command to be run by the next sh invocation below`\
     | sort --unique --field-separator=" " --key=3,3  `# Remove duplicate patch commands to the same source file` \
     | sh `# Execute patch commands`
```

I'm not sure of a workaround that still enables inline comments in long pipelines like above.

Moving all the comments to above the command would remove the valuable context of which part they're referring to.

**Comment by mvdan** [comment]
> I'm not sure of a workaround that still enables inline comments in long pipelines like above.

Indeed, this problem does not have a good solution that I'm aware of. This does not mean that we need to implement one of the bad solutions, though.

Ultimately, such comments are at best a clever hack, and at worst an expensive use of a deprecated feature. The stackoverflow link you pasted goes a bit into detail about that. You're right that shell not having inline comments in long multi-line commands is unfortunate, but that's just how the language is.

shfmt is a formatter, so consistency and sane defaults are very important. The printer does not use backticks anywhere, ever - so adding a special case would hurt consistency, and it would encourage the use of this syntax hack when it shouldn't be.

I think your best bet is to figure out a reasonable way to comment your code without using this hack. https://github.com/mvdan/sh/issues/342#issuecomment-538565468 is one such example. Declaring your arguments in a bash array, which can have comments very easily, is also an option.

**Comment by improbable-nickkrempel** [comment]
> I think your best bet is to figure out a reasonable way to comment your code without using this hack. [#342 (comment)](https://github.com/mvdan/sh/issues/342#issuecomment-538565468) is one such example. Declaring your arguments in a bash array, which can have comments very easily, is also an option.

I don't know of a workaround when the command is a multi-stage pipeline like in the example above.

As a separate issue, do you think the parsing of backticks should be fixed so that at least shfmt doesn't turn working scripts into non-working ones?

**Comment by mvdan** [comment]
> I don't know of a workaround when the command is a multi-stage pipeline like in the example above.

You do have some options, but they involve writing the program in a different way. You have to choose which of the options you prefer; each can have its benefit.

> As a separate but related issue, do you think the parsing of backticks should be fixed so that at least `shfmt` doesn't turn working scripts into non-working ones?

shfmt breaking a program when formatting it is pretty much always a bug :) We can repurpose this issue to fix this case. The bare minimum would be to at least error out when parsing such inline comments, to not end up with an invalid program. But the ideal would be to write a program that follows shfmt's formatting and is still valid.

**Comment by mvdan** [comment]
Ended up implementing this in the PR above; reviews and testing welcome :)

**Comment by MaxWinterstein** [comment]
To shy to create an own issue, but shouldn't this work?

```bash
$ shfmt -version
v3.4.3
$ bat test.sh   
───────┬─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: test.sh
───────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ #!/bin/bash
   2   │ 
   3   │ echo `# this could be any comment you like` \
   4   │ 123
───────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
$ shfmt test.sh 
test.sh:3:6: reached EOF without closing quote `
$ ./test.sh 
123
```

**Comment by mvdan** [comment]
@MaxWinterstein the fix was https://github.com/mvdan/sh/commit/b7f83dc122bab372d3c3945f5c4c30227e6af3a5 in master, which will be included in the upcoming v3.5.0, but wasn't backported to v3.4.x as it's a relatively risky new feature to backport into a bugfix release.

It's been a long time since v3.4.0 though, so I'll look to do another feature release very soon.
